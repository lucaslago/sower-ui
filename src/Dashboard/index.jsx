import React, { Component } from 'react';
import { Row, Col } from 'react-bootstrap';
import DashboardItem from '../DashboardItem';
import AsyncContainer from '../components/AsyncContainer'
import ServerError from '../components/ServerError';

const loadedStyle = {
  marginTop: '10rem',
};

const addSimulationStatus = (device, simulationStatus) => (
  Object.assign({ simulationStatus }, device));

const loadingStyle = {
  textAlign: 'center',
  marginTop: '30rem',
};

const errorStyle = loadingStyle;

export default class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      devices: [],
      dashboardStyle: loadingStyle,
      errorMessage: "We're sorry, something went wrong in our servers",
    };

    this.fetchDevices = this.fetchDevices.bind(this);
    this.renderDashBoardItems = this.renderDashBoardItems.bind(this);
    this.renderServerError = this.renderServerError.bind(this);
  }

  fetchDevices() {
    return this.getDevices()
      .then(response => response.data.map(device => this.getSimulationStatus(device.id)
      .then(status => addSimulationStatus(device, status))))
      .then(devicesPromise => Promise.all(devicesPromise))
      .then((transformedDevices) => {
        this.setState({
          devices: transformedDevices,
          dashboardStyle: loadedStyle,
        });
      })
      .catch((err) => {
        this.setState({ dashboardStyle: errorStyle });
        throw err;
      });
  }

  getSimulationStatus(trackerId) {
    const authToken = this.props.route.authService.getToken();
    return this.props.route.simulationService.status({ trackerId, authToken });
  }

  getDevices() {
    const authToken = this.props.route.authService.getToken();
    return this.props.route.devicesService.fetch(authToken);
  }

  renderDashBoardItems() {
    return this.state.devices.map(device => (
      <Row key={device.id}>
        <DashboardItem
          device={device}
          authToken={this.props.route.authService.getToken()}
          simulationService={this.props.route.simulationService}
        />
      </Row>
      ));
  }

  renderServerError() {
    return (<ServerError message={this.state.errorMessage} />);
  }

  render() {
    return (
      <div style={this.state.dashboardStyle} className="Dashboard">
        <Row>
          <Col xs={12} smOffset={1} sm={10}>
            <AsyncContainer
              promise={this.fetchDevices}
              renderOnResolve={this.renderDashBoardItems}
              renderOnReject={this.renderServerError}
              spinnerSize={80}
              spinnerThickness={5}
            />
          </Col>
        </Row>
      </div>
    );
  }
}

Dashboard.propTypes = {
  route: React.PropTypes.shape({
    authService: React.PropTypes.shape({
      getToken: React.PropTypes.func.isRequired,
    }),
    simulationService: React.PropTypes.shape({
      start: React.PropTypes.func.isRequired,
      stop: React.PropTypes.func.isRequired,
      status: React.PropTypes.func.isRequired,
      create: React.PropTypes.func.isRequired,
    }),
    devicesService: React.PropTypes.shape({
      fetch: React.PropTypes.func.isRequired,
    }),
  }),
};
