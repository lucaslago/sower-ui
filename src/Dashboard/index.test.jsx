import Dashboard from './index';

describe('<Dashboard />', () => {
  const devices = {
    data: [
      {
        type: 'devices',
        id: 'ffcb5837-d34c-48cc-ac76-58fe52c367d4',
        relationships: {
          equipment: {
            data: {
              description: 'A beautiful machine',
              type: 'equipment',
              id: 'b5a48553-0d98-4cbd-b362-cc86e52f1b3e',
            },
          },
        },
      },
    ],
  };
  const authServiceFake = {
    login: jest.fn(),
    loggedIn: jest.fn(),
    getToken: jest.fn(),
    logout: jest.fn()
  };
  const simulationServiceFake = {
    status: jest.fn().mockReturnValue(Promise.resolve({status: 'active', totalPositions: 1000, remainingPositions: 500})),
    stop: jest.fn(),
    start: jest.fn()
  };
  const devicesServiceFake = { fetch: jest.fn().mockReturnValueOnce(Promise.resolve(devices)) };
  const route = {
    authService: authServiceFake,
    devicesService: devicesServiceFake,
    simulationService: simulationServiceFake
  };

  it('renders without crashing', () => {
    const wrapper = shallow(<Dashboard route={route} />);
    expect(wrapper.length).toEqual(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
