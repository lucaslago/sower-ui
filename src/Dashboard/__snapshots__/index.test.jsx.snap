exports[`<Dashboard /> renders without crashing 1`] = `
<div
  className="Dashboard">
  <h1>
     Your devices: 
  </h1>
  <ul />
</div>
`;
